proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,252 @@
+
+The new interface instructions are summarized below.
+
+```
+#interface-instr ::=
+  | bool.lift_iMM
+  | bool.lower_iMM
+
+  | sNN.lift_iMM
+  | sNN.lower_iMM
+
+  | uNN.lift_iMM
+  | uNN.lower_iMM
+
+  | string.lift_memory ..
+  | string.lower_memory ..
+
+  | record.lift ..
+  | record.lower ..
+
+  | array.lift_memory ..
+  | array.lower_memory ..
+
+  | invoke ..
+
+where:
+  NN is one of [8, 16, 32, 64]
+  MM is one of [32, 64]
+```
+
+In addition the following instructions are included from the core instructions:
+
+```
+#interface-instr ::=
+  | ..
+  | <numeric instructions>
+  | <memory instructions>
+  | <table instructions>
+  | drop
+  | unreachable
+  | nop
+```
+
+Refer to the definition of instruction classes [here](https://webassembly.github.io/spec/core/syntax/instructions.html).
+
+## Laziness
+
+Interface values generated by interface instructions are lazy. The execution of an instruction generating an interface value happens immediately before the instruction that consumes the value.
+
+## `bool.lift_iMM`
+
+This instruction has type `[iMM] -> [bool]`.
+
+The output is `true` iff the input `iMM` has any bit set in the integer.
+
+## `bool.lower_iMM`
+
+This instruction has type `[bool] -> iMM`.
+
+The output is `0b1` zero extended to the desired bitwidth iff the input is true.
+
+## `sNN.lift_iMM`
+
+This instruction has type `iMM` -> `sNN`.
+
+The output is the value of `iMM` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `sNN.lower_iMM`
+
+This instruction has type `[sNN] -> [iMM]`
+
+The output is the value of `sNN` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lift_iMM`
+
+This instruction has type `iMM` -> `uNN`.
+
+The output is the value of `iMM` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lower_iMM`
+
+This instruction has type `[uNN] -> [iMM]`
+
+The output is the value of `uNN` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `string.lift_memory`
+
+The full syntax of this instruction is:
+```
+string.lift_memory #memidx #encoding
+```
+
+The type of this instruction is:
+```
+[$base: i32, $len: i32] -> [string]
+```
+
+The output is a sequence of unicode code points taken by interpreting `store.memory[$memidx].bytes[$base .. $base + $len]` with `#encoding`.
+
+This instructions traps if the range `$base .. $base + $len` is out-of-bounds, or the byte sequence is not valid under `#encoding`.
+
+The syntax for supported encodings is:
+```
+#encoding ::=
+  | utf8
+```
+
+Currently, only UTF8 is supported as an encoding. UTF16 may be added in the future.
+
+## `string.lower_memory`
+
+The full syntax of this instruction is:
+```
+string.lower_memory #memidx #encoding $alloc: #funcidx
+```
```

Comment:
currently not anticipating including an allocator with lowering. The complexities of lowering would make this too entangled.

Reply:
also, not all allocations can be traced to a regular core wasm call. May need to be able to bind lambdas.

Reply:
I included allocation to align with how the 'Explainer' describes the string instructions. What complexities does this introduce?



> also, not all allocations can be traced to a regular core wasm call. May need to be able to bind lambdas.

Could you elaborate?

Reply:
memory allocation will very likely be split off from lifting/lowering. This is what Luke and I are currently focusing on.

Reply:
I've split off memory allocation from lowering.